### PR TITLE
Adding abstraction for injecting your own http agent for Android

### PR DESCRIPTION
**Pull request by logangirvin**
**Problem:**
Currently Wallet library has a hard dependency on OkHttpClient, which may not work for every developer. 

**Solution:**
Add an abstraction that can be injected into the wallet library for http agents which can implement the basic request/response needs without much overhead from the developer.

**Validation:**
Initially tested through integration.

**Type of change:**
- [x] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [x] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:

**Documentation Links**:

### CODE REVIEW

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IResponse.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.microsoft.walletlibrary.util.http.httpagent
+
+import java.nio.ByteBuffer
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IResponse.kt]: nit: unused import

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgent.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            }
+
+            in 400..499 -> {
+                Result.failure(ClientError(abstractResponse))
+            }
+
+            in 500..599 -> {
+                Result.failure(ServerError(abstractResponse))
+            }
+
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgent.kt]: nit: extra lines.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgent.kt [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            else -> {
+                Result.failure(UnknownError())
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgent.kt]: Can response body be sent even during known error to have insight into what went wrong rather than empty error state.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/URLFormEncoding.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package com.microsoft.walletlibrary.util.http
+
+import java.io.Serializable
+import java.net.URLEncoder
+
+/**
+ *
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/URLFormEncoding.kt]: nit: missing class comments

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/URLFormEncoding.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+     * @throws InvalidURLEncodeData if any other type is in the Map
+     */
+    fun encode(data: Map<String, Any?>): ByteArray {
+        val builder = StringBuilder()
+        var firstKeyPair = true
+        data.forEach { (key, value) ->
+            val encodedKey = URLEncoder.encode(key, "UTF-8")
+            when (value) {
+                is String -> {
+                    if (!firstKeyPair) {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/URLFormEncoding.kt:23]: nit: move the code that actually performs URL encoding of string to a separate method.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentApiProvider.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import com.microsoft.walletlibrary.util.http.httpagent.IHttpAgent
+import kotlinx.serialization.json.Json
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+internal class HttpAgentApiProvider @Inject constructor(@Named("agent") agent: IHttpAgent, httpAgentUtils: HttpAgentUtils, json: Json) {
+    // maybe refactor these into different interface apis?
+    val presentationApis = HttpAgentPresentationApis(agent, httpAgentUtils)
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentApiProvider.kt:13]: Doesn't presentation response need Json to deserialize the generic response?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentApiProvider.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.microsoft.walletlibrary.did.sdk.util.HttpAgentUtils
+import com.microsoft.walletlibrary.util.http.httpagent.IHttpAgent
+import kotlinx.serialization.json.Json
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+internal class HttpAgentApiProvider @Inject constructor(@Named("agent") agent: IHttpAgent, httpAgentUtils: HttpAgentUtils, json: Json) {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentApiProvider.kt:11]: nit: Add class comments to describe briefly about the functions performed by it.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentIdentifierApi.kt [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import kotlinx.serialization.json.Json
+
+internal class HttpAgentIdentifierApi(private val agent: IHttpAgent,
+                                      private val httpAgentUtils: HttpAgentUtils,
+                                      private val json: Json) {
+
+    fun toIdentifierResponse(response: IResponse): IdentifierResponse {
+        return json.decodeFromString(IdentifierResponse.serializer(), response.body.decodeToString())
+    }
+    suspend fun resolveIdentifier(overrideUrl: String): Result<IResponse> {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentIdentifierApi.kt:21]: nit: no new line.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentIdentifierApi.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package com.microsoft.walletlibrary.did.sdk.datasource.network.apis
+
+import com.microsoft.walletlibrary.did.sdk.identifier.models.identifierdocument.IdentifierResponse
+import com.microsoft.walletlibrary.did.sdk.util.HttpAgentUtils
+import com.microsoft.walletlibrary.util.http.httpagent.IHttpAgent
+import com.microsoft.walletlibrary.util.http.httpagent.IResponse
+import kotlinx.serialization.json.Json
+
+internal class HttpAgentIdentifierApi(private val agent: IHttpAgent,
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentIdentifierApi.kt:13]: Same as above on class comments.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentIssuanceApi.kt [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import kotlinx.serialization.json.Json
+
+internal class HttpAgentIssuanceApi(private val agent: IHttpAgent,
+                                    private val httpAgentUtils: HttpAgentUtils,
+                                    private val json :Json) {
+
+    fun parseContract(response: IResponse): ContractServiceResponse {
+        return json.decodeFromString(ContractServiceResponse.serializer(), response.body.decodeToString())
+    }
+    suspend fun getContract(overrideUrl: String): Result<IResponse> {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentIssuanceApi.kt:22]: nit: new line.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            "vp_token" to vpToken,
+            "state" to state
+        ))
+        return agent.post(
+            overrideUrl,
+            httpAgentUtils.defaultHeaders(HttpAgentUtils.ContentType.UrlFormEncoded, body),
+            body
+        )
+    }
+
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt]: nit: extra line

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt [hunk 12]
```diff
@@ -0,0 +1,55 @@
+package com.microsoft.walletlibrary.did.sdk.datasource.network.apis
+
+import com.microsoft.walletlibrary.did.sdk.util.Constants
+import com.microsoft.walletlibrary.did.sdk.util.HttpAgentUtils
+import com.microsoft.walletlibrary.util.http.URLFormEncoding
+import com.microsoft.walletlibrary.util.http.httpagent.IHttpAgent
+import com.microsoft.walletlibrary.util.http.httpagent.IResponse
+
+internal class HttpAgentPresentationApis(private val agent: IHttpAgent, private val httpAgentUtils: HttpAgentUtils) {
+
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt:14]: Isn't a method needed to convert generic response to Presentation specific ones?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        state: String?
+    ): Result<IResponse> {
+        val body = URLFormEncoding.encode(mapOf<String, Any?>(
+            "id_token" to token,
+            "vp_token" to vpToken,
+            "state" to state
+        ))
+        return agent.post(
+            overrideUrl,
+            httpAgentUtils.defaultHeaders(HttpAgentUtils.ContentType.UrlFormEncoded, body),
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt:36]: Does the body need to be sent in header as well? Wouldn't setting the content-type to application/x-www-form-urlencoded and payload in body be enough?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt:54]: same as above.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt (hunk 13 above)
Comment (nithyaganeshng) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/apis/HttpAgentPresentationApis.kt:36]: I just reviewed the files below. okay so it is only sent to compute the length.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IResponse.kt [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package com.microsoft.walletlibrary.util.http.httpagent
+
+class IResponse(
+    val status: UInt,
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IResponse.kt]: why does the status code have to be UInt and not Int?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IHttpAgent.kt [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package com.microsoft.walletlibrary.util.http.httpagent
+
+abstract class IHttpAgent {
+
+    open class HttpAgentError(val response: IResponse): Error() {}
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IHttpAgent.kt]: Why is this an error and not an exception?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/controlflow/Result.kt [hunk 17]
```diff
@@ -15,6 +16,31 @@ internal sealed class Result<out S> {
...
+        this.getOrNull()?.let {
+            return Result.Success<T>(it)
+        }
+    }
+    this.exceptionOrNull()?.let {
+        if (it is SdkException) {
+            return Result.Failure(it)
+        }
+    }
+    return Result.Failure(SdkException("Could not cast failure to SDK"))
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/controlflow/Result.kt]: Could we append the cause from the exception here so that if it is a failure from a dependency that is bubbled up, it would help in debugging the issue?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/MeasureTime.kt [hunk 18]
```diff
@@ -25,24 +26,39 @@ internal inline fun <R> logTime(name: String, block: () -> R): R {
...
+    var code = 0
+    var cvResponse = "none"
+    var requestId = "none"
+    val result = block().onSuccess {
+        cvResponse = it.headers[Constants.CORRELATION_VECTOR_HEADER] ?: "none"
+        requestId = it.headers[Constants.REQUEST_ID_HEADER] ?: "none"
+        code = it.status.toInt()
+    }.onFailure {
+        when(it) {
+            is IHttpAgent.ClientError -> {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/MeasureTime.kt]: Does the Client and Server error behave differently here? Shouldn't this be a generic HttpAgentError where all of them collect the same information from the response?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/identifier/resolvers/Resolver.kt [hunk 19]
```diff
@@ -18,13 +16,12 @@ internal class Resolver @Inject constructor(
...
-                is Result.Success -> {
-                    Result.Success(id.payload.didDocument)
-                }
-                is Result.Failure -> Result.Failure(ResolverException("Unable to resolve identifier $identifier", id.payload))
+        return identifierRepository.resolveIdentifier(baseUrl, identifier)
+            .map {
+                it.didDocument
+            }
+            .onFailure {
+                return Result.failure(ResolverException("Unable to resolve identifier $identifier"))
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/identifier/resolvers/Resolver.kt]: Can we populate the cause here along with the message for debugging the specific failure cause?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/HttpAgentUtils.kt [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+import com.microsoft.walletlibrary.util.http.URLFormEncoding
+import javax.inject.Inject
+import javax.inject.Named
+
+/**
+ * Internal Class to apply common headers
+ */
+internal class HttpAgentUtils @Inject constructor(@Named("userAgentInfo") private val userAgentInfo: String,
+                                                  @Named("walletLibraryVersionInfo") private val walletLibraryVersionInfo: String,
+                                                  private val correlationVectorService: CorrelationVectorService) {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/HttpAgentUtils.kt:15]: It doesn't use the CorrelationVectorService. correlation vector is currently created for every network call using interceptor s. Doesn't it have to done here?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt [hunk 21]
```diff
@@ -29,52 +30,56 @@ import java.io.IOException
...
-        @Suppress("UNCHECKED_CAST")
-        val transformedPayload = (response.body() ?: throw LocalNetworkException("Body of Response is null.")) as T
-        return Result.Success(transformedPayload)
+        return Result.failure(SdkException("Failed to get a response"))
     }
 
     // TODO("what do we want our base to look like")
-    open fun onFailure(response: Response<S>): Result<Nothing> {
-        val responseBody = response.errorBody()?.string() ?: ""
+    open fun onFailure(exception: Throwable): Result<T> {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt]: For failure cases, it wouldn't return <T> rather a failure result. Was this changed for a specific reason?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt [hunk 22]
```diff
@@ -29,52 +30,56 @@ import java.io.IOException
  * In default methods, S == T, for no transformation takes place.
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt]: nit: can you remove S from class comments since they are obsolete.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/FetchContractNetworkOperation.kt [hunk 23]
```diff
@@ -6,28 +6,26 @@
...
 import com.microsoft.walletlibrary.did.sdk.util.controlflow.InvalidSignatureException
-import com.microsoft.walletlibrary.did.sdk.util.controlflow.IssuanceException
-import com.microsoft.walletlibrary.did.sdk.util.controlflow.Result
+import com.microsoft.walletlibrary.util.http.httpagent.IResponse
 import kotlinx.serialization.json.Json
-import retrofit2.Response
 
 internal class FetchContractNetworkOperation(
     val url: String,
-    apiProvider: ApiProvider,
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/FetchContractNetworkOperation.kt:23]: I have left the comment here, since I couldn't leave it on those classes. Could we remove the old api interfaces and classes since they aren't used anymore?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt [hunk 24]
```diff
@@ -69,12 +72,15 @@ internal object VerifiableCredentialSdk {
         polymorphicJsonSerializers: SerializersModule = Json.serializersModule,
         registrationUrl: String = "",
         resolverUrl: String = "https://discover.did.msidentity.com/v1.0/identifiers",
-        walletLibraryVersionInfo: String = ""
+        walletLibraryVersionInfo: String = "",
+        httpAgent: IHttpAgent = OkHttpAgent()
     ) {
+        correlationVectorService = CorrelationVectorService( PreferenceManager.getDefaultSharedPreferences(context) )
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt:78]: CorrelationVectorService is being injected already below along with other services. Is this necessary here?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/SendIssuanceCompletionResponse.kt [hunk 25]
```diff
@@ -6,18 +6,17 @@
...
-    apiProvider: ApiProvider
-) : PostNetworkOperation<Unit, Unit>() {
-    override val call: suspend () -> Response<Unit> = { apiProvider.issuanceApis.sendCompletionResponse(url, serializedResponse) }
+    apiProvider: HttpAgentApiProvider
+) : PostNetworkOperation<Unit>() {
 
-    override suspend fun onSuccess(response: Response<Unit>): Result<Unit> {
-        return Result.Success(Unit)
+    override val call: suspend () -> Result<IResponse> = { apiProvider.issuanceApis.sendCompletionResponse(url, serializedResponse) }
+    override suspend fun toResult(response: IResponse): Result<Unit> {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/SendIssuanceCompletionResponse.kt:19]: nit: no new line.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/SendPresentationResponseNetworkOperation.kt [hunk 26]
```diff
@@ -6,27 +6,27 @@
...
     }
 }
 
 // The plural vp_token format
-internal class SendPresentationResponsesNetworkOperation(url: String, serializedIdToken: String, vpToken: List<String>, state: String?, apiProvider: ApiProvider) :
-    PostNetworkOperation<String, Unit>() {
-    override val call: suspend () -> Response<String> = {
+internal class SendPresentationResponsesNetworkOperation(url: String, serializedIdToken: String, vpToken: List<String>, state: String?, apiProvider: HttpAgentApiProvider) :
+    PostNetworkOperation<Unit>() {
+    override val call: suspend () -> Result<IResponse> = {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/SendPresentationResponseNetworkOperation.kt:25]: nit: no new line.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/SendVerifiableCredentialIssuanceRequestNetworkOperation.kt [hunk 27]
```diff
@@ -7,59 +7,56 @@ package com.microsoft.walletlibrary.did.sdk.datasource.network.credentialOperati
...
     private val jwtValidator: JwtValidator,
     private val serializer: Json
-) : PostNetworkOperation<IssuanceServiceResponse, VerifiableCredential>() {
-    override val call: suspend () -> Response<IssuanceServiceResponse> = { apiProvider.issuanceApis.sendResponse(url, serializedResponse) }
-
-    override suspend fun onSuccess(response: Response<IssuanceServiceResponse>): Result<VerifiableCredential> {
-        val jwsTokenString = response.body()?.vc ?: throw IssuanceException("No Verifiable Credential in Body.")
+) : PostNetworkOperation<VerifiableCredential>() {
+    override val call: suspend () -> Result<IResponse> = { apiProvider.issuanceApis.sendResponse(url, serializedResponse) }
+    override suspend fun toResult(response: IResponse): Result<VerifiableCredential> {
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/credentialOperations/SendVerifiableCredentialIssuanceRequestNetworkOperation.kt:29]: nit: no new line.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt [hunk 28]
```diff
@@ -7,47 +7,53 @@ import com.microsoft.walletlibrary.did.sdk.credential.service.models.linkedDomai
...
-            verifyLinkedDomains(domainUrls, relyingPartyDid)
-        }
+        getLinkedDomainsFromDid(relyingPartyDid)
+            .onSuccess { domainUrls ->
+                return verifyLinkedDomains(domainUrls, relyingPartyDid)
+            }
+            .onFailure {
+                return Result.failure(it)
+            }
+        return Result.failure(SdkException(""))
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt]: Add an exception message that unknown failure happened while fetching/verifying linked domains.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt [hunk 29]
```diff
@@ -7,47 +7,53 @@ import com.microsoft.walletlibrary.did.sdk.credential.service.models.linkedDomai
...
-                val wellKnownConfigDocument = wellKnownConfigDocumentResult.payload
-                wellKnownConfigDocument.linkedDids.forEach { linkedDidJwt ->
+        if (domainUrls.isEmpty())
+            return Result.success(LinkedDomainMissing)
+        val domainUrl = domainUrls.first()
+        val hostname = URL(domainUrl).host
+        return getWellKnownConfigDocument(domainUrl)
+            .map {
+                wellKnownConfigDocument ->
+                wellKnownConfigDocument.linkedDids.firstNotNullOf { linkedDidJwt ->
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt:45]: Why is only the first processed now instead of all the linked domains in the document?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt [hunk 30]
```diff
@@ -7,47 +7,53 @@ import com.microsoft.walletlibrary.did.sdk.credential.service.models.linkedDomai
...
                     val isDomainLinked = jwtDomainLinkageCredentialValidator.validate(linkedDidJwt, relyingPartyDid, domainUrl)
-                    if (isDomainLinked) return@runResultTry Result.Success(LinkedDomainVerified(hostname))
+                    if (isDomainLinked)
+                        LinkedDomainVerified(hostname)
+                    else
+                        null
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt:50]: If the linked domain being processed is not verified, shouldn't next one be processed instead of returning null?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/URLFormEncoding.kt [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+                }
+                else -> {
+                    throw InvalidURLEncodeData(key)
+                }
+            }
+        }
+        return builder.toString().toByteArray()
+    }
+
+    class InvalidURLEncodeData(val keyName: String) : Error()
```
Comment (nithyaganeshng) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/URLFormEncoding.kt]: Why is this error a first class Error instance and not a Wallet Library Exception?

File: walletlibrary/src/test/java/com/microsoft/walletlibrary/util/http/URLFormEncodingTest.kt [hunk 32]
```diff
@@ -0,0 +1,65 @@
+package com.microsoft.walletlibrary.util.http
+
+import assertk.assertThat
+import assertk.assertions.isEqualTo
+import assertk.assertions.isNotEmpty
+import assertk.fail
+import org.junit.Test
+
+internal class URLFormEncodingTest {
+    @Test
```
Comment (nithyaganeshng) [inline walletlibrary/src/test/java/com/microsoft/walletlibrary/util/http/URLFormEncodingTest.kt:10]: Could you add a test for null string value?

File: walletlibrary/src/test/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgentTest.kt [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+
+            assertThat(request.captured.url.toString()).isEqualTo(expectedURL)
+            assertThat(response.isSuccess).isTrue
+            val unwrapped = response.getOrThrow()
+            assertThat(unwrapped.status).isEqualTo(expectedStatusCode.toUInt())
+            assertThat(unwrapped.body).isEqualTo(expectedPayload)
+            assertThat(unwrapped.headers).isEqualTo(expectedHeaders)
+        }
+    }
+    @Test
```
Comment (nithyaganeshng) [inline walletlibrary/src/test/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgentTest.kt:68]: nit: no new line.

File: walletlibrary/src/test/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgentTest.kt [hunk 34]
```diff
@@ -0,0 +1,108 @@
...
+class OkHttpAgentTest {
+    private val clientMock = mockk<OkHttpClient>();
+    val client = OkHttpAgent();
+
+    @Before
+    fun beforeAllTests() {
+        client.client = clientMock;
+    }
+
+    @Test
```
Comment (nithyaganeshng) [inline walletlibrary/src/test/java/com/microsoft/walletlibrary/util/http/httpagent/OkHttpAgentTest.kt:25]: Could you add one more test for appending headers to default headers and make sure all of them are present and not overwritten?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/HttpAgentUtils.kt (hunk 20 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/util/HttpAgentUtils.kt:15]: This should have been called in the defaultHeaders call

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt (hunk 24 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt:78]: dependency injection uses this sdk as a source for the correlation vector to construct the httpagentutil.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IResponse.kt (hunk 15 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IResponse.kt]: No particular reason. Status codes are all unsigned so I went with that but I think it causes more issues with existing http clients and can change.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IHttpAgent.kt (hunk 16 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/util/http/httpagent/IHttpAgent.kt]: No particular reason; I don't think I was particularly aware of Kotlin having both.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt (hunk 21 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt]: Nope. It also just calls .failure so it could remain Result<Nothing>.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt (hunk 29 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/LinkedDomainsService.kt:45]: I think this was my confusion when coming from Swift. I thought `firstNotNullOf` referred to the returned value of the code block/predicate. I think thats also what Kotlin has? https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/first-not-null-of.html

**Review by nithyaganeshng — APPROVED** [review]
